lib/store/index.js

```diff
@@ -3,29 +3,35 @@
 import { CompositeDisposable } from "atom";
 import { observable, computed, action } from "mobx";
 import { isMultilanguageGrammar, getEmbeddedScope } from "./../utils";
+import _ from "lodash";
 
 import Config from "./../config";
 import MarkerStore from "./markers";
 import kernelManager from "./../kernel-manager";
-import type Kernel from "./../kernel";
+import Kernel from "./../kernel";
+
+import type { IObservableArray } from "mobx";
 
 class Store {
   subscriptions = new CompositeDisposable();
   markers = new MarkerStore();
+  runningKernels: IObservableArray<Kernel> = observable([]);
+  @observable kernelMapping: KernelMapping = new Map();
   @observable startingKernels: Map<string, boolean> = new Map();
-  @observable runningKernels: Map<string, Kernel> = new Map();
   @observable editor = atom.workspace.getActiveTextEditor();
   @observable grammar: ?atom$Grammar;
 
   @computed
   get kernel(): ?Kernel {
-    for (let kernel of this.runningKernels.values()) {
-      const kernelSpec = kernel.kernelSpec;
-      if (kernelManager.kernelSpecProvidesGrammar(kernelSpec, this.grammar)) {
-        return kernel;
-      }
-    }
-    return null;
+    if (!this.filePath) return null;
+    const kernel = this.kernelMapping.get(this.filePath);
+    if (!kernel || kernel instanceof Kernel) return kernel;
+    if (this.grammar) return kernel[this.grammar.name];
+  }
+
+  @computed
+  get filePath(): ?string {
+    return this.editor ? this.editor.getPath() : null;
```

Comment:
We should add a way to handle file renames. If we dont, a rename will put a kernel in limbo.

There is an `onDidChangePath` subscription in the atom api. I think we could use that to call an action here similar to how the editor is updated.

Reply:
Good idea. We can use the editor id for unnamed files then, since it's unlikely that they are closed and re-opend.

> If we don't, a rename will put a kernel in limbo.

That's why we have the Kernel Monitor to handle manage all running kernels.

Reply:
Unfortunately it's not as easy as I thought: `onDidChangePath` only calls back the new path, but we also need the old path in order to correctly change the kernel mapping.

@BenRussert Should we then still allow kernels for unsaved files?